### DISCUSSION TITLE
BPF size limitation

### DISCUSSION THREAD

**Discussion by strasharo** [ZerBea/hcxdumptool#489] [General]
Is there any size limitation for the maximum size of the bpf used for filtering? At 1291 characters it appears to work fine. At 1365 characters it fails with:

scan frequencies: frequency [channel] of Regulatory Domain: 00
  2412 [  1]      2437 [  6]      2462 [ 11]
failed to open raw packet socket
[?25h1 ERROR(s) during runtime
Possible reasons:
 driver is broken
 driver is busy (misconfigured system, other services access the INTERFACE)
1 Packet(s) captured by kernel
0 Packet(s) dropped by kernel
Warning: too less packets received (monitor mode may not work as expected)
Possible reasons:
 driver is broken (most likely)
 no transmitter in range
 frames are filtered out by BPF
1 SHB written to pcapng dumpfile
1 IDB written to pcapng dumpfile
1 ECB written to pcapng dumpfile
0 EPB written to pcapng dumpfile

It's not driver specific, fails on all with the same error. Took me a while to realize the reason, I thought I broke something with the drivers. That's on OpenWRT latest compiled from source.

**Comment by ZerBea** [comment]
The maximum number of instructions (BPF_MAXINSNS) that the unprivileged BPF program can have is limited to 4096.
[Q: What are the verifier limits?](https://docs.kernel.org/6.0/bpf/bpf_design_QA.html#id13)

Both tools check the length:
https://github.com/ZerBea/hcxdumptool/blob/master/hcxdumptool.c#L4440
https://github.com/ZerBea/wifi_laboratory/blob/main/hcxlabtool.c#L4709

If that failed, both tools show an error:
fprintf(stderr, "failed to read BPF\n");

Instrucitons is not equal to characters. FIltereing a single MAC and setting the snaplen are 16 instructions:
```
$ hcxdumptool --bpfc="not wlan addr3 111111111111" | wc
     16      64     157
```
The more MAC to be filtered, the instructions increase:
```
$ hcxdumptool --bpfc="not wlan addr3 111111111111 and not wlan addr3 222222222222" | wc
     19      76     196

$ hcxdumptool --bpfc="not wlan addr3 111111111111 and not wlan addr1 222222222222" | wc
     20      80     204

$ hcxdumptool --bpfc="not wlan addr1 111111111111 and not wlan addr2 222222222222" | wc
     26     104     267
```

**Reply by ZerBea** [reply]
I just made an attempt.
This code generates a command line of 501 MACs to be filtered:
```
#define _GNU_SOURCE
#include <ctype.h>
#include <dirent.h>
#include <fcntl.h>
#include <ftw.h>
#include <libgen.h>
#include <stdarg.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <utime.h>
#include <inttypes.h>

/*===========================================================================*/
int main()
{
int a = 0;

uint64_t mac = 0x112233445566;

fprintf(stdout,"hcxdumptool --bpfc=\"not wlan addr3 112233445566");
for(a = 0; a < 500; a++)
	fprintf(stdout, " and not wlan addr3 %012" PRIx64, mac + a);
fprintf(stdout,"\"\n");
return EXIT_SUCCESS;
}
```

It compiles as expected (hcxdumptool/hcxlabtool) and the size of the BPF is 760 instructions:
```
$ wc test.bpf
  760  3040 11492 test.bpf
```

Driver is rtl8xxxu:
```
$ hcxdumptool -l
  10	 13	5ce931984b44	980ee4b8cd3b	+	wlp5s0f3u2      	rtl8xxxu	NETLINK
```

Everything is as expected:
```
$ sudo hcxdumptool -i wlp5s0f3u2 --rcascan=active --bpf=test.bpf
...
^C
265 Packet(s) captured by kernel
0 Packet(s) dropped by kernel
77 PROBERESPONSE(s) captured
```

How many instructions does your bpf contain?

**Comment by strasharo** [comment]
This is the sample bpf which fails it for me when being ran (last three octets swapped with FF just for the sample):

`not wlan addr1 D8:07:B6:FF:FF:FF and not wlan addr2 D8:07:B6:FF:FF:FF and not wlan addr3 D8:07:B6:FF:FF:FF and not wlan addr1 AC:8B:A9:FF:FF:FF and not wlan addr2 AC:8B:A9:FF:FF:FF and not wlan addr3 AC:8B:A9:FF:FF:FF and not wlan addr1 AC:8B:A9:FF:FF:FF and not wlan addr2 AC:8B:A9:FF:FF:FF and not wlan addr3 AC:8B:A9:FF:FF:FF and not wlan addr1 B2:8B:A9:FF:FF:FF and not wlan addr2 B2:8B:A9:FF:FF:FF and not wlan addr3 B2:8B:A9:FF:FF:FF and not wlan addr1 AC:15:A2:FF:FF:FF and not wlan addr2 AC:15:A2:FF:FF:FF and not wlan addr3 AC:15:A2:FF:FF:FF and not wlan addr1 F0:B4:29:FF:FF:FF and not wlan addr2 F0:B4:29:FF:FF:FF and not wlan addr3 F0:B4:29:FF:FF:FF and not wlan addr1 A4:7E:39:FF:FF:FF and not wlan addr2 A4:7E:39:FF:FF:FF and not wlan addr3 A4:7E:39:FF:FF:FF and not wlan addr1 30:05:05:FF:FF:FF and not wlan addr2 30:05:05:FF:FF:FF and not wlan addr1 DC:4F:22:FF:FF:FF and not wlan addr2 DC:4F:22:FF:FF:FF and not wlan addr1 5C:CF:7F:FF:FF:FF and not wlan addr2 5C:CF:7F:FF:FF:FF and not wlan addr1 64:90:C1:FF:FF:FF and not wlan addr2 64:90:C1:FF:FF:FF and not wlan addr1 5C:E5:0C:FF:FF:FF and not wlan addr2 5C:E5:0C:FF:FF:FF and not wlan addr1 8C:C8:4B:FF:FF:FF and not wlan addr2 8C:C8:4B:FF:FF:FF and not wlan addr1 70:03:9F:FF:FF:FF and not wlan addr2 70:03:9F:FF:FF:FF and not wlan addr1 00:E0:4C:FF:FF:FF and not wlan addr2 00:E0:4C:FF:FF:FF`
That would be 37 instructions I guess.

**Reply by ZerBea** [reply]
I'm still hunting for the problem. The filter compiles fine and it works as expected. Driver is rtl8xxxxu.
```
$ sudo hcxdumptool -i wlp5s0f3u3 --rcascan=active --bpf=test.bpf
...
506 Packet(s) captured by kernel
0 Packet(s) dropped by kernel
109 PROBERESPONSE(s) captured

exit on sigterm
```

```
$ wc test.bpf
 319 1276 3590 test.bpf
```
319 - last new line == 318 instrutcions which is also far below the limits.

This are my limits:
```
$ xargs --show-limits
Your environment variables take up 1434 bytes
POSIX upper limit on argument length (this system): 2093670
POSIX smallest allowable upper limit on argument length (all systems): 4096
Maximum length of command we could actually use: 2092236
Size of command buffer we are actually using: 131072
Maximum parallelism (--max-procs must be no greater): 2147483647

Execution of xargs will continue now, and it will try to read its input and run commands; if this is not what you wanted to happen, please type the end-of-file keystroke.
Warning: echo will be run at least once.  If you do not want that to happen, then press the interrupt keystroke.
^C
```
The command line length to build the filter is far below this limits.

**Reply by strasharo** [reply]
Just to note I'm running the latest package version on OpenWRT which is 6.3.5 (I've modified the Makefile locally to pick it up). Building the bpf works fine, however it fails on runnning.
I'll now build the latest source from git, just to confirm if it exist there as well. Platform is RPI Zero W, latest openwrt-24.10 built from source.

**Reply by ZerBea** [reply]
You'll run into the same problem.
It looks like there is a limitation that my system doesn't have.

**Reply by ZerBea** [reply]
The problem is located here:
https://github.com/ZerBea/hcxdumptool/blob/master/hcxdumptool.c#L4079

I'll add some debug code.

**Reply by strasharo** [reply]
Tried to run xargs --show-limits on OpenWRT, but it doesn't have that argument there. 

```
  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 24.10-SNAPSHOT, r28218-45bf30e062
 -----------------------------------------------------
root@zerowrt:~# xargs --show-limits
xargs: unrecognized option: show-limits
BusyBox v1.36.1 (2024-12-24 10:25:39 UTC) multi-call binary.

Usage: xargs [OPTIONS] [PROG ARGS]

Run PROG on every item given by stdin

	-0	NUL terminated input
	-o	Reopen stdin as /dev/tty
	-r	Don't run command if input is empty
	-t	Print the command on stderr before execution
	-p	Ask user whether to run each command
	-E STR,-e[STR]	STR stops input processing
	-n N	Pass no more than N args to PROG
	-s N	Pass command line of no more than N bytes
	-x	Exit if size is exceeded
root@zerowrt:~# 

```

**Reply by ZerBea** [reply]
OK, its not compiled in. No problem.

This version included a debug output:
[hcxdumptool.c.zip](https://github.com/user-attachments/files/18245407/hcxdumptool.c.zip)
replace hcxdumptool.c by this version and recompile.
Than run it to retrieve a detailed ERROR message (errno).

This are the changes:
```
if(bpf.len > 0)
	{
	int retdump = 0;
	retdump = setsockopt(fd_socket_rx, SOL_SOCKET, SO_ATTACH_FILTER, &bpf, sizeof(bpf));
	fprintf(stdout, "SO_ATTACH_FILTER failed: %s\n", strerror(errno));
	if(retdump < 0) return false;
	}
```

**Reply by ZerBea** [reply]
Improved the default debug mode of hcxdumptool and hcxlabtool by latest commits.
Now SO_ATTACH_FILTER has been added to the debug log, if something went wrong here.

**Reply by ZerBea** [reply]
Instead of switching the source code, you can enable it the debug log in Makefile::
```
# uncomment to enable DEBUG log
#DEFS		+= -DHCXDEBUG
```

**Reply by strasharo** [reply]
Sorry for the delay, it took me a while to figure out the Makefile for the feeds package, but here it is with the latest HEAD:
```

root@zerowrt:~/dump# cat hcxdumptool.log 
SO_ATTACH_FILTER failed: Out of memory
root@zerowrt:~/dump# free -m
              total        used        free      shared  buff/cache   available
Mem:         443824       14156      411844         132       17824      396116
Swap:             0           0           0
root@zerowrt:~/dump# ulimit
unlimited
root@zerowrt:~/dump# 
```

**Reply by ZerBea** [reply]
Thanks. The issue is not related to you memory.

It could be possible that the MAX_USED_MAPS limit have been exceeded (default == 64) and the verifier rejects the BPF without providing any extra log messages:
https://github.com/torvalds/linux/blob/8bb7eca972ad531c9b149c0a51ab43a417385813/kernel/bpf/verifier.c#L11629-L11632

At the moment, I have no idea why some systems running out of memory.

What happens if you try to capture traffic running tcpdump or tshark with this filter applied?

**Reply by ZerBea** [reply]
Looks like MAX_USED_MAPS is really hard coded to 64:
https://github.com/torvalds/linux/blob/9b2ffa6148b1e4468d08f7e0e7e371c43cac9ffe/include/linux/bpf_verifier.h#L589

That means, a filter can have 4096 instructions, but if it is too complex (e.g. many loops), MAX_USED_MAPS limit exceeds.

The only thing I can do is to print an ERROR message if this happens.

**Reply by ZerBea** [reply]
But you can reduce complexity of your filter code.
An example:

code is not optimized: addr3, addr2, addr3, addr3
```
$ hcxdumptool --bpfc="not wlan addr3 111111111111 and not wlan addr2 444444444444 and not wlan addr3 333333333333 and not wlan addr3 222222222222" | wc
     36     144     391
```

optimzed code to reduce loops: addr3, addr3, addr3, addr2
```
$ hcxdumptool --bpfc="not wlan addr3 111111111111 and not wlan addr3 222222222222 and not wlan addr3 333333333333 and not wlan addr2 444444444444" | wc
     32     128     353
```
Less instructions and less loops.

**Reply by strasharo** [reply]
Thanks a lot for the clarification. Just to confirm that I'm understanding the filtering right - APs appear as addr1,addr2 and addr,3
Clients appear as addr1 and addr2. Is that correct?
And sorting them by address type reduces the loops?

**Reply by ZerBea** [reply]
BTW:
_APs appear as addr1,addr2 and addr3. Clients appear as addr1 and addr2._
It depends on the frame types:
- frames only with addr1, e.g. ACK
- frames only with addr1 and addr2, e.g. REQUEST_TO_SEND
- frames with addr1, addr2 and addr3, e.g. BEACON
- frames with addr1, addr2, addr3 and addr4, e.g. mesh data frames
https://howiwifi.com/2020/07/13/802-11-frame-types-and-formats/

hcxdumptool/hcxlabtool only works on frames which are mandatory to retrieve a hash. All other frames are ignored (due to performance reason).

If you want to filter an AP it is enough to filter ADDR3.

If you want to filter a CLIENT it is enough to filter ADDR2. This will fail if the CLIENT is running MAC randomization.
hcxdumptool/hcxlabtool runs MAC randomization. If the CLIENT runs MAC randomization too, the filter can't prevent that we get an EAPOL M1M2ROGUE.

To improve your filter, I suggest to use tshark (live capturing):
```
$ sudo hcxdumptool -m INTERFACE_NAME -c 1a
$ tshark -i INTERFACE_NAME (for unfiltered traffic)
$ tshark -i INTERFACE_NAME -f "....." (for filtered traffic), e.g. -f "wlan addr3 111111111111"
```

BTW:
I found an interesting description here (especially the MAPS section):
https://docs.cilium.io/en/latest/reference-guides/bpf/architecture/#maps
_The maximum instruction limit per program is restricted to 4096 BPF instructions, which, by design, means that any program will terminate quickly. For kernel newer than 5.1 this limit was lifted to 1 million BPF instructions._

And in the MAPS section:
_A single BPF program can currently access up to 64 different maps directly._
It looks like your filter exceeds this limit and the kernel reports ERRNO "out of memory".

**Reply by strasharo** [reply]
Thanks a lot for the detailed explanation. Something's weird though, it started failing with just one statement with the latest build:

```
root@zerowrt:~/dump# service hcxdump start 
Sleeping for 10 seconds before starting the service...
Starting service with retry interval 10 and max retries 5.
PHY phy0 already has an interface. Skipping.
PHY phy1 already has an interface. Skipping.
Wireless device found: archer
Preparing BPF filter.
Generated BPF filter: not wlan addr3 D8:07:B6:XX:XX:XX 
Starting /usr/sbin/hcxdumptool on device archer.
root@zerowrt:~/dump# cat hcxdumptool.log 
SO_ATTACH_FILTER failed: No error information
root@zerowrt:~/dump# 
```
It's running fine without a BPF set.

**Reply by ZerBea** [reply]
Thanks. That should be fixed now (by latest commit).
The debug code (to hunt for the problem) has now been completely removed. Attaching a BPF always caused a message and hcxdumptool terminates regardless if it is an ERROR or not.

**Reply by strasharo** [reply]
Thanks a lot, it works perfectly now. With your guidance I've reduced my bpf by a few times. Here's the routine is use to generate it in my init script in case someone might find it useful to borrow. It's looking for both ap/client black/whitelists and generates a bpf with them if they're being present. 

```
    echo "Preparing BPF filter."

    # Step 1: Remove the old filter file
    /bin/rm -f /root/dump/filter.bpf

    # Step 2: Initialize BPF filter variable
    bpf_filter=""

    # Step 3: Process blacklist_ap if available
    if [ -f /root/dump/blacklist_ap ]; then
        while read -r line; do
            mac_address=$(echo "$line" | awk '{print $1}')
            if [ -n "$mac_address" ]; then
                bpf_filter="${bpf_filter}not wlan addr3 $mac_address and "
            fi
        done < /root/dump/blacklist_ap
    fi

    # Step 4: Process blacklist_client if available
    if [ -f /root/dump/blacklist_client ]; then
        while read -r line; do
            mac_address=$(echo "$line" | awk '{print $1}')
            if [ -n "$mac_address" ]; then
                bpf_filter="${bpf_filter}not wlan addr2 $mac_address and "
            fi
        done < /root/dump/blacklist_client
    fi

    # Step 5: Process target_ap if available
    if [ -f /root/dump/target_ap ]; then
        while read -r line; do
            mac_address=$(echo "$line" | awk '{print $1}')
            if [ -n "$mac_address" ]; then
                bpf_filter="${bpf_filter}wlan addr3 $mac_address or "
            fi
        done < /root/dump/target_ap
    fi

    # Step 6: Process target_client if available
    if [ -f /root/dump/target_client ]; then
        while read -r line; do
            mac_address=$(echo "$line" | awk '{print $1}')
            if [ -n "$mac_address" ]; then
                bpf_filter="${bpf_filter}wlan addr2 $mac_address or "
            fi
        done < /root/dump/target_client
    fi

    # Remove trailing "and " or "or " from the BPF filter
    bpf_filter=${bpf_filter%and }
    bpf_filter=${bpf_filter%or }

    # Step 7: Generate filter.bpf if BPF filter is non-empty
    use_bpf_arg=""
    if [ -n "$bpf_filter" ]; then
        $PROG --bpfc="$bpf_filter" > /root/dump/filter.bpf
        echo "Generated BPF filter: $bpf_filter"
        use_bpf_arg="--bpf=/root/dump/filter.bpf"
    else
        echo "No filters generated. Running without BPF filter."
    fi

    # Step 8: Start hcxdumptool
    echo "Starting $PROG on device $WLANDEV."
    /usr/sbin/screen -AmdS dump -L -Logfile /root/dump/screen.out \
        $PROG $use_bpf_arg -i "$WLANDEV" -w /root/dump/"$ARCHIVENAME".pcapng --rds=1
```

**Reply by ZerBea** [reply]
You're welcome. Glad to hear that it is working now and thanks for sharing your script. 

Sorting by MAC ADDR field reduce the list of instructions and loops.

BTW:
To reduce workload, some control frame types (type ctl / subtype ack and type ctl / subtype cts) can be filtered out too:
`$ hcxdumptool --bpfc="not type ctl subtype ack and not type ctl subtype cts not wlan addr3 D807B6FFFFFF and not wlan addr3 ..."`
This prevents that this frames types reach hcxdumptools/hcxlabtools man packet loop.
